Mission service - add section on mission execution

hamishwillee opened the pull request:
Recently had a query "how do I pause a mission". Made it clear to me that we don't have a clear explanation of the execution cycle of a mission - start, pause, reset.

@peterbarker @julianoes @tridge Would appreciate your review. I think this is accurate based on previous discussions with @tridge . For example, it captures (without being specific) that ArduPlane switches out of Mission mode when the mission is complete, while other vehicles (Copter say) loiter within the mission mode.

julianoes commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
+### Pausing/Stopping Missions
+
+Pause and stop missions by changing mode.
+In MAVLink this can be done using [MAV_CMD_DO_SET_MODE](../messages/common.md#MAV_CMD_DO_SET_MODE) (generally a vehicle will also support changing modes via an RC controller/joystick too).
```
E.g. to Hold, at least for PX4.

hamishwillee replied:
Any mode change will do it. I didn't mention a specific mode because there's also some differences in Copter vs Fixed wing with related to what happens if you touch sticks in a mission (Fixed wing does not switch to Hold)

julianoes commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
+Once a mission has started it will iterate through the mission items, jumping and looping as indicated by jump mission items, until it completes.
+While the mission is running you can change the current mission item, but this does not reset the jump counters used for loops, and once the mission is complete, changing the current mission item will not restart it.
+
+The mission will reset after you land and the vehicle disarms.
```
Are you sure?

hamishwillee replied:
I thought so. What did you think - only on reboot?

julianoes replied:
On reboot for sure.

julianoes replied:
And on re-upload, or manual reset to 0.

peterbarker replied:
ArduPlane will not reset counters if you land the vehicle in auto.  It's a bit more complicated because some missions have the vehicle land and disarm, but then be able to resume the mission from where they are in their sequence when they're armed in auto again.

hamishwillee replied:
> And on re-upload, or manual reset to 0.

@julianoes Just to be clear, 
- I understand reupload on the ground will reset the counters and the current waypoint. 
- For upload in air the current waypoint does not change right? On PX4 you "finish the previously uploaded waypoint" and then the waypoint number increments and you progress to that waypoint in the newly uploaded mission.

> On manual reset to 0

What do you mean by this? Is this the same as ArduPilot - setting mission current to 0 resets the whole mission? If so, that would be good. 

> ArduPlane will not reset counters if you land the vehicle in auto. It's a bit more complicated because some missions have the vehicle land and disarm, but then be able to resume the mission from where they are in their sequence when they're armed in auto again.

@peterbarker Thanks very much. This caveat is true of all PX4 vehicle too - its a configuration setting or a mission type setting. I will add a comment.

julianoes replied:
> For upload in air the current waypoint does not change right? On PX4 you "finish the previously uploaded waypoint" and then the waypoint number increments and you progress to that waypoint in the newly uploaded mission.

This is ground station dependent. I think QGC does that. But if you just upload the whole mission, it will go back to 0.

> setting mission current to 0 resets the whole mission? If so, that would be good.

I think so, except things like DO_JUMP counters.

hamishwillee replied:
@julianoes 

> I think so, except things like DO_JUMP counters.

So it will remove the "complete" bit and restart the mission, but with the wrong counters? That's screwed up. Can we make it also reset the counters? (i.e. as part of this protocol) and make not resetting them non-standard.

FYI also @peterbarker my concern here is that we updated [MAV_CMD_DO_SET_MISSION_CURRENT](https://mavlink.io/en/messages/common.html#MAV_CMD_DO_SET_MISSION_CURRENT) to add a reset option because there was not way to reset the counters. Now it seems like setting the mission item to 0 would do this. That's a better approach. So we should standardize.
I.e. I think we'd be best off removing that separate parameter and stating that setting to zero should reset the counters and the complete bit and reset to zero. Thoughts?

julianoes replied:
> So it will remove the "complete" bit and restart the mission, but with the wrong counters? That's screwed up. Can we make it also reset the counters? (i.e. as part of this protocol) and make not resetting them non-standard.

That's the way I made it in 2013 or so. It might have been fixed since. You could try it out...

hamishwillee replied:
I doubt it. In any case, Peter commented somewhere here he prefers that counters are reset explicitly. So we probably need to implement that second parameter. I looked at it once but couldn't work out the right place to even add the command.

peterbarker commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
+In either case, the mission cannot be restarted unless it is [reset](#resetting-missions) (which usually happens after the vehicle lands and disarms).
+
+While the mission is executing the current mission item can be changed via MAVLink ([MAV_CMD_DO_SET_MISSION_CURRENT](#MAV_CMD_DO_SET_MISSION_CURRENT)).
+Note however that this does not reset the jump counters, and if the mission is complete, will not restart the mission.
```
Setting the waypoint to zero in ArduPilot will reset the counters.

ArduCopter (and some other ArduPilot vehicles) reset the mission upon disarm.

hamishwillee replied:
I'll move all the caveats into the reset the mission section so we can handle in one place.

```suggestion
Note however that unless the command [resets the mission](#resetting-missions), this does not reset the jump counters, and if the mission is complete, will not restart the mission.
```

peterbarker commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
+If supported the [MAV_CMD_MISSION_START](../messages/common.md#MAV_CMD_MISSION_START) command will both arm the vehicle (if necessary) and switch to the mission mode.
+This command also allows you to specify a particular start and end item in the mission.
+
+### Pausing/Stopping Missions
```
I find it *very* strange that there's no coverage of `MAV_CMD_DO_PAUSE_CONTINUE` here!  That's the more appropriate malvink-ish way of doing this behaviour, I'd think.

Saying "do-set-mode" won't always stop the vehicle from executing waypoints - for example, ArduPlane when you've got a `DO_LAND_START`.  If you switch modes to RTL it's essentially the same thing as changing waypoints - the Plane will jump to its landing sequence.

hamishwillee replied:
> I find it very strange that there's no coverage of MAV_CMD_DO_PAUSE_CONTINUE here! That's the more appropriate malvink-ish way of doing this behaviour, I'd think.

Doh. That's my brain fade. Thanks. Will fix.

> Saying "do-set-mode" won't always stop the vehicle from executing waypoints - for example, ArduPlane when you've got a DO_LAND_START. If you switch modes to RTL it's essentially the same thing as changing waypoints - the Plane will jump to its landing sequence.

@peterbarker This is true on PX4 too but there is a nuance that waypoints are being executed but you're still in Return mode, not mission mode. Is it the same on ArduPlane?

The TLDR here is that PX4 RTL has settings to execute a return mode landing following a mission landing pattern. So you are following waypoints, but you're not in a mission.

peterbarker replied:
> @peterbarker This is true on PX4 too but there is a nuance that waypoints are being executed but you're still in Return mode, not mission mode. Is it the same on ArduPlane?

I believe so.  Unless behaviour has changed the vehicle will briefly appear to go into RTL mode but will then appear to go back to AUTO mode.

hamishwillee replied:
@peterbarker What if you're not in auto mode to start with - will it go to RTL mode and still then go to AUTO mode to do the landing? That seems odd if the intent was to RTL.

hamishwillee replied:
In any case, fixed.

peterbarker commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
+### Resetting Missions
+
+Once a mission has started it will iterate through the mission items, jumping and looping as indicated by jump mission items, until it completes.
+While the mission is running you can change the current mission item, but this does not reset the jump counters used for loops, and once the mission is complete, changing the current mission item will not restart it.
```
As mentioned, changing to wp 0 in ArduPilot will reset.

hamishwillee commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
 
+## Executing Missions
+
+Following mission upload, or vehicle restart, or mission reset, all jump counter loops are set to their initial values.
```
Hmmm. This doc treats full mission upload as "standard" and partial as non-standard. I don't really agree with that, but it makes no difference here, so ...

```suggestion
Following full mission upload, or vehicle restart, or mission reset, all jump counter loops are set to their initial values.
```

hamishwillee commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
+## Executing Missions
+
+Following mission upload, or vehicle restart, or mission reset, all jump counter loops are set to their initial values.
+Unless a mission has just been uploaded that explicitly sets the current mission item, the current mission item is set to 1.
```
@peterbarker Yes I do. Thanks for clarification. There is a further exception here that if you are flying the mission item is not reset. That allows uploading a new mission and it continuing on the same mission item.
```suggestion
The current mission item is normally reset to one on mission reset or vehicle restart, or on mission upload when the vehicle is landed. The current mission item may retain its current value if a mission is uploaded in-air or it may be set to the mission item that has `current` set in a mission item (this field is ignored in ArduPilot but respected by PX4).
```

hamishwillee commented on en/services/mission.md:
```diff
@@ -379,6 +379,51 @@ Note:
+Following mission upload, or vehicle restart, or mission reset, all jump counter loops are set to their initial values.
+Unless a mission has just been uploaded that explicitly sets the current mission item, the current mission item is set to 1.
+
+The mission is executed when the vehicle is armed and in the flight stack's mission mode.
```
@peterbarker I'll change that to be a minimum requirement.

```suggestion
For a mission to be executed the vehicle must be armed and in the flight stack's mission mode.
There may (or may not) be further requirements to start a mission.
For example PX4 will start as soon as you arm in mission mode, while ArduCopter in its default configuration requires that the throttle is raised.
Similarly, planes that are configured for throttle/hand launch may require a minimum acceleration before the mission is executed. 
```

hamishwillee commented on en/services/mission.md:
```diff
@@ -379,6 +379,55 @@ Note:
+
+On completion of all mission items the mission will "complete".
+At this point some flight stacks will loiter within the mission mode, while others will transition to some other mode (such as "Hold") and not allow you to return to mission mode.
+In either case, the mission cannot be restarted unless it is [reset](#resetting-missions) (which usually happens after the vehicle lands and disarms).
```
```suggestion
In either case, the mission cannot be restarted unless it is [reset](#resetting-missions).
```

hamishwillee commented:
@peterbarker @julianoes OK, I have overhauled this. I think I've addressed all points.

This does some additional restructure to fully separate mission upload/download from mission execution. I've reduced the initial description of mission execution so most of the detail is in the sub-sections. The caveats are mostly in the resetting section.

peterbarker commented:
On Tue, 5 Mar 2024, Hamish Willee wrote:

> @peterbarker What if you're not in auto mode to start with - will it go to RTL mode and still then go to AUTO mode to do the landing? That seems odd if the intent was to RTL.

Yeah.  Coincidentally I'm playing with this at the moment.

We have a parameter "RTL_AUTOLAND" which determines the behaviour.  0 
disables the autoland behaviour, it just goes to home or a rally point.

A value of 1 send you home (you will report as RTL), and then send you to 
the landing sequence (will then report as auto).

A value of 2 will send you straight to the landing sequence.  You will 
*briefly* (one heartbeat!) report as RTL then report as auto.

peterbarker commented:
On Tue, 5 Mar 2024, Hamish Willee wrote:

> FYI also @peterbarker my concern here is that we updated MAV_CMD_DO_SET_MISSION_CURRENT to add a reset option because there was not way to reset the counters. Now it seems like setting the mission item to 0
> would do this. That's a better approach. So we should standardize.

No!  I don't think that will work for PX4 because that's a *realy 
waypoint* for PX4, not the home position like it is in ArduPilot!

i.e. you might have a legitimate reason in PX4 to set the waypoint to zero 
while preserving your loop counters.

> I.e. I think we'd be best off removing that separate parameter and stating that setting to zero should reset the counters and the complete bit and reset to zero. Thoughts?

I think we should stop relying on the 
magic-waypoint-0-resets-loop-counters behaviour in the ground control 
stations, and use that new parameter or some other mechanism.  I don't 
think I'm all that keen on that new parameter, honestly.  Does it mean you 
can't reset the loop counters without also possibly changing your 
waypoint?

Note that ArduPilot does not currently honour that new parameter.

peterbarker commented:
> Does it mean you can't reset the loop counters without also possibly changing your waypoint?

...reads documentation.  Never mind....

hamishwillee commented:
> Note that ArduPilot does not currently honour that new parameter.

Neither does PX4. It's on my list. I'd just like consistency.

hamishwillee commented:
> A value of 2 will send you straight to the landing sequence. You will *briefly* (one heartbeat!) report as RTL then report as auto.

Thanks. I guess its just a bit different than I am used to. On PX4 the same behaviours exist, but the thinking is that whether or not you're executing a mission landing, if you're RTL due to a failsafe or whatever that is the mode of interest for users.
I think I agree - I'd be surprised to do RTL and suddenly see myself in a mission.

Anway, have a look at the update. Let me know how you go with reseting loop counters.
